### ISSUE TITLE
Make weapon slots consistent (3 weapon system revisited)

### ISSUE THREAD

**Issue by squeek502** [fortressforever/fortressforever#93]
Split from #82, where Largos said:

> We could break every class down to "Primary, Secondary, Tertiary/Melee"
> 
> Scout - Nail, shotty, crowbar
> Sniper - Sniper, AR, crowbar (nailgun seems totally worthless on sniper too)
> Soldier - RPG, Shotgun, melee
> Demo - Exception. Two slots for two types of pipes, then shotgun, then crowbar, then detpack
> HWGuy - AC, shotgun, melee
> Medic - Supernail, shotgun, medkit
> Spy - tranq, shotgun, knife
> Engie - Rail, shotgun, wrench

This idea has been discussed a lot in the past. Some quotes from the dev forums in 2009:

squeek:

> The main point is to not restrict what each class can do, while still clarifying and simplifying the usage of the class. 

---

squeek:

> This whole 3 weapon thing is hinging on how we address the demo, imo. If we just remove his single shotty, he'll lose depth/options/usefulness (blue pipe + single shotty combo) and people will not be happy, as it is only a restrictive change. If we give him some sort of hitscan, we have two options: 1) altfire key allowing combination of blue/grellow pipe launchers or blue/hitscan or 2) some sort of combination of the blue pipe launcher/hitscan that doesn't require an altfire button.

---

CrazyCarl:

> Take away the crowbar and give [the demoman] a pea shooter with unlimited ammo. 

---

AfterShock:

> this is what we'd be removing:
> 
>    Scout: Nothing
>     Sniper: Nothing
>     Soldier: single shotgun (he has a super shotgun)
>     Demoman: ?
>     Medic: single shotgun (he has a super shotgun)
>     HWGuy: single shotgun (he has a super shotgun.. and his AC is hitscan anyway)
>     Pyro: single shotgun (could argue this but i can't see why i'd ever want the pyro playing with a shotgun when he has a unique burn multiplier system)
>     Spy: Nailgun (should be sabotaging not nailing SGs)
>     Engineer: Nothing
> 
> The other alternative to simply removing the spy nailgun is to allow the tranq some sort of SG effect. Personally I'd like to improve the spy as an all-round SG opponent and reduce the overwhelming power of the sabotage to compensate. Perhaps if you shoot the SG with the tranq, it goes offline for x seconds or something (like an EMP shot). Would give you a window to kill the engineer and disguise before then sabotaging, might be a nice change to constantly walking around cloaked..
## My current thoughts

This is a good idea, and I think removing the demo's crowbar and replacing it with the single shotgun (or a separate demo-specific weapon that serves the same purpose) would be fine. Unsure of which order the weapons should go (slot1 melee or slot3 melee?).

We could also supplement this with a `cl_tfcweaponslots` cvar that remaps the weapon slots to their TFC equivalents if people don't want to adjust.

**Comment by squeek502** [comment]
An idea for combining the blue pipe launcher with the functionality of the single shotgun from that dev forum thread:

squeek:

> blue pipe launcher, if you hold down +attack, after a second or so it shoots a hitscan burst

---

squeek:

> A short charge time for the blue pipe launcher to shoot a hitscan burst is appealing to me. It would allow the demo to choose between a blue pipe or a hitscan blast each time he attacks, and would create the same depth that the blue pipe/single shotty choice created.

---

GreenMushy:

> To expand on what squeek said about the demo launcher, cuz its a cool idea: The premise was to keep the functionality of what the singly shotty did by combining it with the blue launcher. The blue launcher could shoot hitscan out of it by holding the button down for a very short window, the pipe explodes in the barrel, an shoots out hitscan. For one, it kindof makes sense, and it maintains what the demoman needs: a quick hitscan source. We get the 3 weapon system we want, and we dont restrict what the demoman can do. If were feeling risky, we can also then separate the ammo for the blue and yellow launcher, which is something i would be interested in trying. So the class could still blue pipe + quick hitscan combo by pressing then holding. Ideally, the demoman should just have yellow and blue launchers, and that would really accentuate what the class is about. Although, i do find that the demo does find very cool uses for having hitscan, so id hate to see that go. But with this solution, we get the best of both worlds by simplifying it to just 2 launchers and keeping the hitscan. Its a win-win. This is a good example of what ideas should look like, as they all should be "addition without restriction". Or is it "removal without restriction"? Its gotta be one of those.

---

AfterShock:

> My concerns with it are:
> - It's unintuitive for a grenade launcher to fire shotgun pellets if you hold fire down. (can't think of any other games that do this)
> - It's unintuitive to hold down fire to make the weapon do something completely different
> - If holding down fire doesnt affect the normal blues at all then it's even more unintuitive.
> - You would lose the instant fire functionality of the shotgun as you'd need to wait for it to charge up before it fires
> - Changing anything so it has a charge time might make it laggy to use (see the railgun)
> - We'd lose the "hold down fire to fire nonstop blues" functionality

---

GreenMushy:

> I think the idea is that u hold the blue pipe in the gun for .5 seconds, and it blows up inside the gun, shootnig out shrapnel. So, it kindof doe make sense, if the explosions inside the gun sends the message of "something just exploded in my gun". Im no sure saying its unintuitive that holding does something different, because in TF2, holding your sticky launcher starts a charge time. Its not the same thing, but its still the idea that u cant spam it, and holding it does something different then pressing it. And it seems to work fine in that game. Wed have to make the hold as quick as possible to try to replicate the quick switch to shotty + shoot speed. 

It never got tried and it still seems potentially interesting to me.

**Comment by FDA-** [comment]
I agree with Aftershock about demo on all points. I also would like to see the spy keep his nail gun. It's a very strong, if situational team work tool. Just go ahead and remove single shotty from those 3 classes and take nails from sniper, it's not a big change and it's not even directly addressing the problem of class misinterpretation. The idea of the demo pipe/hitscan combo is exceptionally goofy in that regard. It's taking a mechanic that is is potentially misinterpreted by newer players and hiding it away from them. The idea sounds very clunky, and could confuse players into thinking "hold click to charge, charge = power" which brings back the same problem in a new clunky way.

**Comment by ddm999** [comment]
I agree with FDA on this. Making it have 3 weapon slots for all classes just hides away parts from new players rather than simplifying it.

The other issue is that the only way to implement this for Demoman is to simply make a strange and obscure method of firing because him and Engineer are the only classes where basically all their weapons are required to make him the classes they are.

In TF2 Spy and Engy have 4 weapons, and it doesn't make them "more confusing". There's no reason to place a 3 weapon system in FF when 4/5 isn't really even "worse" or "bad".

**Comment by squeek502** [comment]
@FDA-:

> The idea sounds very clunky, and could confuse players into thinking "hold click to charge, charge = power" which brings back the same problem in a new clunky way.

Very true.

> I also would like to see the spy keep his nail gun. It's a very strong, if situational team work tool.

But it muddies the spy's role a bit. Should he stand back and kill an SG from far away or should he use his unique abilities to get close to an SG and kill it?

@ddm999:

> The other issue is that the only way to implement this for Demoman is to simply make a strange and obscure method of firing

This isn't true. A straightforward solution would be to remove the demo's crowbar and keep his single shotty. The idea I put in the second post is just some crazy thing that I thought was kind of cool.

> In TF2 Spy and Engy have 4 weapons, and it doesn't make them "more confusing". There's no reason to place a 3 weapon system in FF when 4/5 isn't really even "worse" or "bad".

3 isn't a magic number, but things being consistent is useful as it allows us to follow a convention for slot numbers like Largos proposed and forces us to cut down on clutter. Either way, though, engi/demo/scout would keep their buildable "weapons".

One thing that I think is being missed is that each weapon that a class has is something of a liability: new players will have to reason about which weapon to use at any given time, and that's made all the more difficult with each additional weapon. Should a brand new player be expected to figure out when the nailgun is a useful weapon for Spy? Would we really lose that much if the nailgun didn't exist on the Spy? Is it worth the extra burden placed on new players?

One thing I've learned is that people can _always_ be worse than you ever could have imagined at making choices about weapons/classes in FF. Every single video of a brand new player I've ever seen struggles with this.

**Comment by FDA-** [comment]
" Should he stand back and kill an SG from far away or should he use his unique abilities to get close to an SG and kill it?"

What ever gets the job done without discrimination. Idk, I see what you're getting at but I'm not sure removing his ng is necessary or for the best.

Also don't take crowbar away from demo. Melee is kind of useless outside of "for the lols" play or an absolute last ditch effort. But taking it away is just trying to force the concept in an entirely futile manner. It's flailing your arms as the ship goes down. Solves nothing other than a totally arbitrary "3 weapon" standard.

Obviously you need to do a proper job of conveying certain concepts, but eventually you need to leave it up to the player and there's no real getting around that.

You and I have shared a few fresh blood videos, but you need to look at them in a different light perhaps. What type of player do you think is making this sort of video? FF is an obscure free source mod. I would wager a lot of these "lost puppy" looking players are younger kids, with no money for other games and nothing else to do, and you could probably watch them have similar experiences in any video game, much less FF. Truth is there's a lot of players playing a lot of games in this fashion it's just much more obvious to you in relation to FF, because you're much more invested in FF, you're looking for this type of thing, and the smaller scale in general.

Rearrange the weapon slots. 1 being your strongest or "primary" and then go down the to the weaker weapons. There is more that could be done in readability, such as models, sounds, hints, and tutorials but I wouldn't worry to much about it honestly. That was rambley and I got lost and ran out of gas.

**Comment by Largos** [comment]
I thank you for using my original post to start this conversation, and if there are some classes that still need 4 slots I say it's okay.  The part that was silly was that soldier has NO slot4 yet his primary is in slot5.  I'm all for going up to 4 or 5 slots as long as the progression of slot1 to slotx is intuitive.  You're on the right track though, it's totally legit to drop single shotty for a lot of classes and nailgun from sniper.

Agreed with FDA about demoman.  Don't try to add a hitscan mechanic with a projectile weapon.

How about the tweak to pipes GreenMush talked about?  

The TF2 demoman can charge his sticky launcher, meaning the longer he holds fire, the further the pipe comes out.  This could be added to BOTH blue and yellow pipes, allowing the demo to make some ridiculous blue pipe airshots while adding depth to his primary weapon.  This falls in the "adding without restriction" category.

As for spy, I agree that the nailgun doesn't quite jive with what the spy could be.  I say replace nailgun with grapplehook, but I"ll keep campaigning for movement techniques til the day FF dies.  (Didn't you enjoy the double jump gun, squeek?)

**Comment by squeek502** [comment]
Note to myself/anyone that works on this: the weapon script system currently locks weapons into specific slots (all single shotguns occupy slot2, not currently possible to have it in a different slot per class). Would need to implement a solution to that in order to start swapping weapon slots freely.

@Largos

> (Didn't you enjoy the double jump gun, squeek?)

[Yes, I did](https://github.com/fortressforever/fortressforever/pull/85). The spy stuff is [sitting in a branch](https://github.com/fortressforever/fortressforever/tree/2.5_spy_patch) ([or two](https://github.com/fortressforever/fortressforever/tree/features/spy-overhaul)) as well. All that stuff wasn't forgotten and will likely be revisited soon enough.

Try to keep discussion on topic, though. Feel free to create new issues for anything you want to discuss.
